### ISSUE TITLE
Random video stream fails to start with "Socket error: Error"

### ISSUE THREAD

**Issue by longzheng** [Sunoo/homebridge-camera-ffmpeg#1087]
**Describe The Problem:**
Randomly through the week, I will not be able to stream the video for any of the cameras for a short duration (a few minutes).

I would see `Socket error: Error` in the logs for any of the cameras I would try to view.

I've ruled out issues with the network or RTSP streams because if I load the RTSP streams manually in VLC, they will all work. Also it would affect all of my cameras pointing to different streams.

After waiting a few minutes, I would be able to successfully view the stream without any other change.

Since the only that differs between my failed and successful attempts are the `rtcpport` or `srtp` ports which look randomized, and the fact that the error is "socket error", I wonder if I'm running into some sort of port conflict with other apps/services running on my server.

**To Reproduce:**
1. Randomly view a camera stream
2. See error

**Logs:**

```
[25/03/2021, 3:08:57 pm] [Camera FFmpeg] [Front] Video stream requested: 1280 x 720, 30 fps, 299 kbps
[25/03/2021, 3:08:57 pm] [Camera FFmpeg] [Front] Starting video stream: 1280 x 720, 30 fps, 299 kbps (AAC-eld)
[25/03/2021, 3:08:57 pm] [Camera FFmpeg] [Front] Stream command: C:\Users\Long\AppData\Roaming\npm\node_modules\homebridge-camera-ffmpeg\node_modules\ffmpeg-for-homebridge\ffmpeg.exe -i rtsp://XXX@192.168.1.4:554/cam/realmonitor?channel=7&subtype=0 -an -sn -dn -codec:v libx264 -pix_fmt yuv420p -color_range mpeg -r 30 -f rawvideo -preset ultrafast -tune zerolatency -filter:v scale='min(1280,iw)':'min(720,ih)':force_original_aspect_ratio=decrease,scale=trunc(iw/2)*2:trunc(ih/2)*2 -b:v 299k -payload_type 99 -ssrc 15449906 -f rtp -srtp_out_suite AES_CM_128_HMAC_SHA1_80 -srtp_out_params JF7l4xudkzUzmenEU8MH4eJmgwuLVj2ft/YqRqD5 srtp://192.168.1.141:58456?rtcpport=58456&pkt_size=1316 -vn -sn -dn -codec:a libfdk_aac -profile:a aac_eld -flags +global_header -f null -ar 16k -b:a 24k -ac 1 -payload_type 110 -ssrc 14230879 -f rtp -srtp_out_suite AES_CM_128_HMAC_SHA1_80 -srtp_out_params o4swsHRBqo9qSPrvFqJMis03SReauio6XlsJQWs+ srtp://192.168.1.141:64227?rtcpport=64227&pkt_size=188 -loglevel level+verbose -progress pipe:1
[25/03/2021, 3:08:57 pm] [Camera FFmpeg] [Front] Socket error: Error
[25/03/2021, 3:08:57 pm] [Camera FFmpeg] [Front] Stopped video stream.
[25/03/2021, 3:08:57 pm] [Camera FFmpeg] [Front] FFmpeg exited with code: null and signal: SIGKILL (Expected)
[25/03/2021, 3:09:06 pm] [homebridge-camera-ffmpeg] This plugin slows down Homebridge. The write handler for the characteristic 'Selected RTP Stream Configuration' on the accessory 'Front F9B7' didn't respond at all!. Please check that you properly call the callback! See https://git.io/JtMGR for more info.
[25/03/2021, 3:09:06 pm] [Camera FFmpeg] [Front] Stopped video stream.
```

Whereas a few seconds later it would work

```
[25/03/2021, 3:09:11 pm] [Camera FFmpeg] [Front] Video stream requested: 1280 x 720, 30 fps, 299 kbps
[25/03/2021, 3:09:11 pm] [Camera FFmpeg] [Front] Starting video stream: 1280 x 720, 30 fps, 299 kbps (AAC-eld)
[25/03/2021, 3:09:11 pm] [Camera FFmpeg] [Front] Stream command: C:\Users\Long\AppData\Roaming\npm\node_modules\homebridge-camera-ffmpeg\node_modules\ffmpeg-for-homebridge\ffmpeg.exe -i rtsp://XXX@192.168.1.4:554/cam/realmonitor?channel=7&subtype=0 -an -sn -dn -codec:v libx264 -pix_fmt yuv420p -color_range mpeg -r 30 -f rawvideo -preset ultrafast -tune zerolatency -filter:v scale='min(1280,iw)':'min(720,ih)':force_original_aspect_ratio=decrease,scale=trunc(iw/2)*2:trunc(ih/2)*2 -b:v 299k -payload_type 99 -ssrc 6255934 -f rtp -srtp_out_suite AES_CM_128_HMAC_SHA1_80 -srtp_out_params pLbX84olcvQ2rYULpJp5d61a/LvAhGz1XNpPrU0g srtp://192.168.1.141:58252?rtcpport=58252&pkt_size=1316 -vn -sn -dn -codec:a libfdk_aac -profile:a aac_eld -flags +global_header -f null -ar 16k -b:a 24k -ac 1 -payload_type 110 -ssrc 2469502 -f rtp -srtp_out_suite AES_CM_128_HMAC_SHA1_80 -srtp_out_params eWXOJ2sWB2C1JZ2YMJ7SXyksedE3dekG1tlEkNTl srtp://192.168.1.141:49300?rtcpport=49300&pkt_size=188 -loglevel level+verbose -progress pipe:1
[25/03/2021, 3:09:11 pm] [Camera FFmpeg] [Front] [info] ffmpeg version N-96777-gebee808595-ffmpeg-windows-build-helpers Copyright (c) 2000-2020 the FFmpeg developers
[25/03/2021, 3:09:11 pm] [Camera FFmpeg] [Front] [info]   built with gcc 8.3.0 (GCC)
[25/03/2021, 3:09:11 pm] [Camera FFmpeg] [Front] [info]   configuration: --pkg-config=pkg-config --pkg-config-flags=--static --extra-version=ffmpeg-windows-build-helpers --enable-version3 --disable-debug --disable-w32threads --arch=x86_64 --target-os=mingw32 --cross-prefix=/ffmpeg-windows-build-helpers/sandbox/cross_compilers/mingw-w64-x86_64/bin/x86_64-w64-mingw32- --enable-libcaca --enable-gray --enable-libtesseract --enable-fontconfig --enable-gmp --enable-gnutls --enable-libass --enable-libbluray --enable-libbs2b --enable-libflite --enable-libfreetype --enable-libfribidi --enable-libgme --enable-libgsm --enable-libilbc --enable-libmodplug --enable-libmp3lame --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libopus --enable-libsnappy --enable-libsoxr --enable-libspeex --enable-libtheora --enable-libtwolame --enable-libvo-amrwbenc --enable-libvorbis --enable-libwebp --enable-libzimg --enable-libzvbi --enable-libmysofa --enable-libopenjpeg --enable-libopenh264 --enable-liblensfun --enable-libvmaf --enable-libsrt --enable-demuxer=dash --enable-libxml2 --enable-opengl --enable-libdav1d --enable-libsvthevc --enable-libaom --enable-libvpx --enable-nvenc --enable-nvdec --extra-libs=-lharfbuzz --extra-libs=-lm --extra-libs=-lpthread --extra-cflags=-DLIBTWOLAME_STATIC --extra-cflags=-DMODPLUG_STATIC --extra-cflags=-DCACA_STATIC --enable-amf --enable-libmfx --enable-gpl --enable-avisynth --enable-frei0r --enable-filter=frei0r --enable-librubberband --enable-libvidstab --enable-libx264 --enable-libx265 --enable-libxvid --enable-libxavs --enable-avresample --extra-cflags='-mtune=generic' --extra-cflags=-O3 --enable-static --disable-shared --prefix=/ffmpeg-windows-build-helpers/sandbox/cross_compilers/mingw-w64-x86_64/x86_64-w64-mingw32 --enable-nonfree --enable-decklink --enable-libfdk-aac
[25/03/2021, 3:09:11 pm] [Camera FFmpeg] [Front] [info]   libavutil      56. 41.100 / 56. 41.100
[25/03/2021, 3:09:11 pm] [Camera FFmpeg] [Front] [info]   libavcodec     58. 70.100 / 58. 70.100
[25/03/2021, 3:09:11 pm] [Camera FFmpeg] [Front] [info]   libavformat    58. 38.101 / 58. 38.101
[25/03/2021, 3:09:11 pm] [Camera FFmpeg] [Front] [info]   libavdevice    58.  9.103 / 58.  9.103
[25/03/2021, 3:09:11 pm] [Camera FFmpeg] [Front] [info]   libavfilter     7. 76.100 /  7. 76.100
[25/03/2021, 3:09:11 pm] [Camera FFmpeg] [Front] [info]   libavresample   4.  0.  0 /  4.  0.  0
[25/03/2021, 3:09:11 pm] [Camera FFmpeg] [Front] [info]   libswscale      5.  6.100 /  5.  6.100
[25/03/2021, 3:09:11 pm] [Camera FFmpeg] [Front] [info]   libswresample   3.  6.100 /  3.  6.100
[25/03/2021, 3:09:11 pm] [Camera FFmpeg] [Front] [info]   libpostproc    55.  6.100 / 55.  6.100
[25/03/2021, 3:09:11 pm] [Camera FFmpeg] [Front] [tcp @ 00000218bb6cdbc0] [verbose] Starting connection attempt to 192.168.1.4 port 554
[25/03/2021, 3:09:11 pm] [Camera FFmpeg] [Front] [tcp @ 00000218bb6cdbc0] [verbose] Successfully connected to 192.168.1.4 port 554
```

**Homebridge Config:**

```json
        {
            "name": "Camera FFmpeg",
            "porthttp": 5005,
            "cameras": [
                {
                    "name": "Backyard",
                    "unbridge": true,
                    "videoConfig": {
                        "source": "-i rtsp://XXX@192.168.1.4:554/cam/realmonitor?channel=2&subtype=0",
                        "audio": true,
                        "debug": true
                    }
                },
                {
                    "name": "Front",
                    "unbridge": true,
                    "videoConfig": {
                        "source": "-i rtsp://XXX@192.168.1.4:554/cam/realmonitor?channel=7&subtype=0",
                        "audio": true,
                        "debug": true
                    }
                },
                {
                    "name": "Intercom",
                    "doorbell": true,
                    "unbridge": true,
                    "videoConfig": {
                        "source": "-i rtsp://192.168.1.3/h264_stream",
                        "stillImageSource": "-i http://192.168.1.3/api/camera/snapshot?width=1280&height=960",
                        "vcodec": "copy",
                        "audio": true,
                        "debug": true
                    }
                },
                // removed other entries for brevity
            ],
            "platform": "Camera-ffmpeg"
        },
```

**Screenshots:**

**Environment:**

* **Node.js Version**: v14.14.0
* **NPM Version**: 6.14.10
* **Homebridge Version**: 1.3.3
* **Homebridge Camera FFmpeg Version**: 3.1.2
* **Homebridge Config UI X Plugin Version**: 4.40.0
* **Operating System**: Windows 10

**Comment by longzheng** [comment]
The code I must be hitting is this https://github.com/Sunoo/homebridge-camera-ffmpeg/blob/9073cf5573808ea0cbba79dfa7005f1884a064ff/src/streamingDelegate.ts#L412

- Is the socket creation random by `dgram`?
- Does `createSocket()` check that it is not created on a used port?
 - If so, is there a bug?
 - If not, does this need a retry behaviour to try different ports for a few attempts?
- Is there any other information in the `error` object that could be logged for better debugging?

**Comment by Sunoo** [comment]
Not sure why you’d be getting access denied on opening a socket, I’ve not come across that before. Is your OS particularly locked down or something?

**Comment by Sunoo** [comment]
A quick search suggests that perhaps Windows firewall could be causing this? I haven’t had a chance to do much testing on Windows, so I can’t say for sure.

**Comment by longzheng** [comment]
I'll look into that, but can you help me understand how the `videoReturnPort` works? https://github.com/Sunoo/homebridge-camera-ffmpeg/blob/master/src/streamingDelegate.ts#L276

Seems to be just a random port right?

**Comment by longzheng** [comment]
I'm running on a pretty standard Windows 10 installation. I do have some other services in the background (Docker, a PBX service) which may be taking ports, but I would have expected `get-port` to check for used ports before it returns one?

**Comment by Sunoo** [comment]
Yea, should be a random, unused, high port. The iOS device sends RTCP packets to that port while the camera is being watched, and watching to see if those packets stop is a way to prevent FFmpeg processes from running forever in cases where the iOS device drops off the network instead of cleanly shutting the stream down.

**Comment by longzheng** [comment]
@Sunoo I wanted to clarify, why does the `audioReturnPort` does not seem to have a binding? Does it not need one? https://github.com/Sunoo/homebridge-camera-ffmpeg/blob/9073cf5573808ea0cbba79dfa7005f1884a064ff/src/streamingDelegate.ts#L278

**Comment by Sunoo** [comment]
The iOS device sends RTCP packets back for the audio stream separately from the ones for the video stream. I don’t currently bother watching the ones from the audio.

**Comment by longzheng** [comment]
I've narrowed down the program a bit more.

I have a PBX server running on the same machine, which by default seems to reserve a large port range (UDP 45000-65000) for VOIP RTP sessions.

![image](https://user-images.githubusercontent.com/484912/112770907-cb848d80-9074-11eb-80a8-4d46952afdbe.png)

For some reason, `get-port` still returns those ports already in use randomly (or supplied as a preferred port).

For example, port `57000` is definitely in use, but it would still return it and then binding would fail with `EACCES`

```
        const port = await getPort({port: 57000});
        console.log(`port selected: ${port}`)
        // port selected: 57000
        const server = dgram.createSocket('udp4');
        server.bind(port);
        // Error: bind EACCES 0.0.0.0:57000
```

I don't know whether this is a bug with `get-port` not checking ports correctly, or something with Windows not allowing it to check correctly.

**Comment by longzheng** [comment]
Hang on a second, I think `get-port` is only for finding free **TCP** ports, whereas you are binding a UDP port?

**Comment by longzheng** [comment]
Yep confirmed that's the problem. `get-port` uses `net.createServer` to create a TCP server on a specific port to test if it works. https://github.com/sindresorhus/get-port/blob/main/index.js#L24-L28

However, a port can be independently used between TCP and UDP, so whilst it might be free for TCP, it might be reserved for UDP.

```
        const tcpServer = net.createServer();
        tcpServer.listen(57000);
        // tcp works

        const udpServer = dgram.createSocket('udp4');
        udpServer.bind(57000);
        // udp fails
```

**Comment by longzheng** [comment]
I wonder if this library will work since it seems to use Sockets to check for ports https://github.com/baalexander/node-portscanner

**Comment by Sunoo** [comment]
Well, that’s embarrassing.

In my defense, I got that code from another developer (not gonna name names though :P ). I guess I’ve just been lucky this whole time.

I’ll try to get that resolved in the next version, if I can find library that does what I actually need. I’ll look into portscanner to see if it fits the bill when I have a second.

**Comment by longzheng** [comment]
After investigating a bit further, it looks like if you just call `socket.bind()` without a port, the OS will automatically find an (open) port.

I made a test case that tries to bind 100,000 times and it succeeded every time so I can confirm this indeed works.

```
const dgram = require("dgram");

async function getTestPort() {
    return new Promise(async (resolve, reject) => {
       
        const server = dgram.createSocket('udp4');
        //server.bind(port);
        server.bind();

        server.on('error', (err) => {
            console.log(`server error:\n${err.stack}`);
            server.close();
            resolve(false);
        });

        server.on('listening', () => {
            const address = server.address();
            console.log(`server listening ${address.address}:${address.port}`);
            server.close();
            resolve(true);
        });
    });
}

(async () => {
    let success = 0;
    let failed = 0;
    for (let index = 0; index < 100000; index++) {
        await getTestPort() ? success++ : failed++;
    }
    console.log(`success: ${success} | failed: ${failed}`)
})();

// success: 100000 | failed: 0
```

**Comment by Sunoo** [comment]
Interesting. The documentation doesn’t mention looking for an open port though:

> If port is not specified or is 0, the operating system will attempt to bind to a random port.

I’ll have to do some poking to see if that works for me too, since I need to know the port number before setting up the socket, so I can tell HomeKit where to send the RTCP packets.

**Comment by longzheng** [comment]
I'm inclined to say the documentation is just inconsistent, because the TCP equivelent does explicitly state the OS will pick an unused port https://nodejs.org/docs/latest/api/net.html#net_server_listen_port_host_backlog_callback
>If port is omitted or is 0, the operating system will assign an arbitrary unused port, which can be retrieved by using server.address().port after the 'listening' event has been emitted.

**Comment by longzheng** [comment]
I wonder if as a quick drop-in replacement you can just swap out `getPort()` with this helper

```javascript
const dgram = require("dgram");

async function getTestPort(): Promise<number> {
    return new Promise(async (resolve, reject) => {
       
        const server = dgram.createSocket('udp4');
        server.bind();

        server.on('error', () => {
            server.close();
            reject();
        });

        server.on('listening', () => {
            const port = server.address().port;
            server.close();
            resolve(port);
        });
    });
}
```

**Comment by Sunoo** [comment]
It could be an issue with the documentation. I intended to actually check the Node source code to confirm the actual functionality, but I didn’t have a chance yet.

**Comment by longzheng** [comment]
I'm curious to learn more about what iOS actually sends through RTCP? What happens if it's not there?

**Comment by Sunoo** [comment]
The reason why I listen for these packets (I do not inspect them or anything, just make sure they’re flowing) is because there are some cases where the iOS device doesn’t tell HomeKit it is done with the camera stream (such as when the device drops off the network) which results in the FFmpeg process running forever, since nothing ever tells it to stop. So I added this functionality to manually tear down the stream when it appears the iOS device has disappeared.

**Comment by Sunoo** [comment]
From what I can see, the documentation stating it uses a random port is misleading. It asks the OS to provide an available port, so it should be fine.

I’ll have a fix for this in the next version, whenever I have a chance to pull that together. Hopefully shouldn’t be too long, but there is another fix that I have mostly in place that I also intend to include in that version.

**Comment by longzheng** [comment]
> From what I can see, the documentation stating it uses a random port is misleading. It asks the OS to provide an available port, so it should be fine.

👍 you should submit a PR to update the nodeJS documentation

> I’ll have a fix for this in the next version, whenever I have a chance to pull that together. Hopefully shouldn’t be too long, but there is another fix that I have mostly in place that I also intend to include in that version.

Do you want me to take a stab at putting together a PR based on my `getPort` alternative?

**Comment by Sunoo** [comment]
I’d hold off on a PR, since I’m thinking it probably makes more sense to make a slightly deeper change than just whipping up a replacement. I’ll give it some thought, but I’m thinking something like along the lines of passing the open socket around instead of opening a socket, getting the port, closing it, and then reopening it.

And yea, I probably will PR their documentation, since the current wording isn’t accurate and causes confusion.

**Comment by longzheng** [comment]
@Sunoo Just wondering if you were still working on this. If you're busy, I could take a stab at it?

**Comment by Sunoo** [comment]
Sorry, moved recently and still settling in. I have a solution half coded based on [pick-port](https://www.npmjs.com/package/pick-port), hoping to have a chance to finish that up soon and get it pushed.

**Comment by longzheng** [comment]
@Sunoo Appreciate you might be busy but let me know if you need help with this as I'm keen to get this fixed.

**Comment by Sunoo** [comment]
Sorry, real life has been absolutely crazy recently. Really hoping to get the new version pushed this week though.

**Comment by longzheng** [comment]
Understand and no worries. Good luck 👍

**Comment by Sunoo** [comment]
Figured I’d check in since stale bot popped back up. This fix should be released tomorrow.

**Comment by Sunoo** [comment]
I lied, this should be fixed in 3.1.3
